### ISSUE TITLE
Confusing compiler message with mapped type 

### ISSUE THREAD

**Issue by Tomassito** [microsoft/TypeScript#46507]
# Bug Report

### 🔎 Search Terms

mapped type incorrect error

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, including nightly and I reviewed the FAQ for entries about mapped type 'bugs which are not bugs'

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.0-dev.20211024#code/KYDwDg9gTgLgBMAdgVwLZwMrLJWwAmAYgJYA2MwUAhgEanADOcA3gFBwdwCCGGAkgHEAcgFERcALxwA5FQYNiAc0TBg0gDTtOAEQCqIgPrauAFXFTp+ZMG1UKGrRxNcBGSTJhVFDBwF9WrKC48EhocCJCJnwmAJoGsQAK4myc3O7SXA6pAELp2X4BQdDwMACeYMBwANLApQxciPgAalSk1gwGXO4pnADaWDjFBCTklLT0DAB0PPzCYgC6cABcLBVQDBCIfPgrKKg0lL6985qp-djBw2QU1HSMk3qGxmaLK8wMnrC2FCsfUMSIRTqBCNb7AX4wf6A3yafyBcDFOBlCrVWr1RotNqMAy5KQ9DjnQZ4IjXMZ3KYzQSiESvVaUDZbHZwPYHKBHE6OOCEy4k0a3CaTZyuWnMTyKbYQqFA5lUVDguB-AGKGGsOHw4JI8qVESIGDEMomLUmCAjG41OoNZqtdrdTm9CJRWLxGJJaa083oq1YjpcO0O6JxRIiSbZd1oy2Y9o41UBfDAADGpCoUEqADNkIh43rNkjGDBCBmADwiEwIEAURpMf1OoMAPgAFJyAWBkPAlvjUiC9QatSsS8DOalU6SoG9egBrWpwAFwSelCCp8K6-WlQ0VY2mygeiPWxj2kzzWk67uro0mkfbjG7hj7+YT2rzfydzi+ACUyzgVEQpQCFA+BcQetmCQE813latAxdYMuGBYc+TeXw3zgAB6ZC4EUJBKGIeM4AAd2gccmAYSB4E2ABCAJWGRSpMWIfAAGEIH2AE7GITYGELABZLgElrW0zhLadEFnWoFzgbiEnmN4QJXMC+xMYE50jeU5zEiTb1gkcGDHOchJE+dF3UktDw-IyD3vUpHzgfwjlUwyePmSi40TZM0wzLM2OExBNgETD-njAD6zAKhSlICAqCZWiGKYmgWOzRAOOPWSz03KBLy9dpa3fFYvx-VhUOnGBpCYCBxzgAALShKmAAA3JAkXKiBkEUcq4Dgm4mF08qAT1QFpyYbz4HwqBxyVVhvMQXyVH8wKOxknsKj7SIA2dV0YMHJTdxWAYeVS8Z7kpOYRFOTt2vpN5B07bkhl5G59opXgqQWFZemYNYGQlOAAEZfA5Z8zh2m69vJB59CMUwaTeD5kxgMEVmkDQQXwOGZGkJ9-o4Xw4Eu1I3wCCapqwgKMyAzl5tPRbwmWmsoOmE7OE2rFtouIGR3u6ZHqO4ECsUCAIHwTVKbs1ELSvb1Okus71gujGCUB4lgYFQ7qSkrk3vpTZPp+v7-oKyqUzgeMmLAMhKAVEKmBgcq7DgIU3FcuBUGIeQlQQOrhKtpqWoaqd8E2aR4AOMK+pgCAkTDiDVug4qOHS5SfWxjGCpoVtJjT6dF3nZA4D9pEoFKcO4DAKAIBqujKit4BUGBI3UBN+goHNuoGpt-UFUa5BSAFwa4AOCqqpxlC0LllmFbZkG7dV0UvE+hHgUQWV5QRrHB8xxPZbfIA)

### 💻 Code

```ts
export enum SupportedFilterables {
    ASSIGNEE = 'assignee',
    DUE_DATE = 'dueDate',
    TAGS = 'tags',
}

export enum ENTITY_TYPE {
    A = 'A',
    B = 'B',
}

export type KeysAndValues_A = {
    [SupportedFilterables.ASSIGNEE] : {personId: number}[],
    [SupportedFilterables.DUE_DATE] : {startDate: string, endDate: string},
}

export type KeysAndValues_B = {
    [SupportedFilterables.ASSIGNEE] : {personId: number}[],
    [SupportedFilterables.TAGS] : {tagId: string, name: string},
}


export type EntityTypeToFilterKeysAndValues = {
    [ENTITY_TYPE.A] : KeysAndValues_A
    [ENTITY_TYPE.B] : KeysAndValues_B
}

type ValidCombinations<MAP> = {
    [ET in keyof MAP]: {entityType: ET, keyValue: keyof MAP[ET], filters: {[key in keyof MAP[ET]] : MAP[ET][key]} }
}[keyof MAP]


declare function nonGenericFun(payload: ValidCombinations<EntityTypeToFilterKeysAndValues>) : any

nonGenericFun({
    entityType: ENTITY_TYPE.A,
     keyValue: SupportedFilterables.ASSIGNEE, // good type: keyof KeysAndValues_A
      filters: {
          [SupportedFilterables.ASSIGNEE]: [{personId: 1}],
        // here compiler says that TAGS are missing even though they don't belong to to ENTITY_TYPE.A's  KeysAndValues_A 
        // but... if you do try to provide them, compiler says that it should not be here
        //   [SupportedFilterables.TAGS]: {tagId: '', name: ''} 
          } 
          })
```

### 🙁 Actual behavior

Compiler getting confused as to what key is missing in 'filters' even though 'keyValue' is evaluated correctly.

### 🙂 Expected behavior

We get correct hints for the filters keys in nonGenericFun. 
The generic one (from sandbox) works but some use cases (e.g. https://github.com/reduxjs/redux-toolkit/issues/1598) require resorting to other ways.

**Comment by tadhgmister** [comment]
Some clarification, the error message is a little misleading but it starts with `Type '{ assignee: { personId: number; }[]; }' is not assignable to type 'KeysAndValues_A | KeysAndValues_B'.` and then typescript demonstrates that it isn't equal for the last case in the union it checked that didn't pass. if you [use the entity type that is the last one in the union](https://www.typescriptlang.org/play?ts=4.5.0-dev.20211024#code/KYDwDg9gTgLgBMAdgVwLZwMrLJWwAmAYgJYA2MwUAhgEanADOcA3gFBwdwCCGGAkgHEAcgFERcALxwA5FQYNiAc0TBg0gDTtOAEQCqIgPrauAFXFTp+ZMG1UKGrRxNcBGSTJhVFDBwF9WrKC48EhocCJCJnwmAJoGsQAK4myc3O7SXA6pAELp2X4BQdDwMACeYMBwANLApQxciPgAalSk1gwGXO4pnADaWDjFBCTklLT0DAB0PPzCYgC6cABcLBVQDBCIfPgrKKg0lL6985qp-djBw2QU1HSMk3qGxmaLK8wMnrC2FCsfUMSIRTqBCNb7AX4wf6A3yafyBcDFOBlCrVWr1RotNqMAy5KQ9DjnQZ4IjXMZ3KYzQSiESvVaUDZbHZwPYHKBHE6OOCEy4k0a3CaTZyuWnMTyKbYQqFA5lUVDguB-AGKGGsOHw4JI8qVESIGDEMomLUmCAjG41OoNZqtdrdTm9CJRWLxGJJaa083oq1YjpcO0O6JxRIiSbZd1oy2Y9o41UBfDAADGpCoUEqADNkIh43rNkjGDBCBmADwiEwIEAURpMf1OoMAPgAFJyAWBkPAlvjUiC9QatSsS8DOalU6SoH3dfrSoaKsbTZQPRHrYx7SZ5oPOL4AJTLOBURClAIUD4FxD15hIbuT3vhSIB52urjA4d8t6+TdwAD077giiQlGI8bgAB3aAAGsmAYSB4E2ABCAJWGRSpMWIfAAGEIH2AE7GITYGELABZLgElrW0zhLOAATgEDaggVM4AIhJ5jec8JyneV+0o2pI3lKjShoujCOXE44CfG4GBWejBP8I4eL4+jV1jBMkxTYSMyzbDEGZTYBF-f542PeswCoUpSAgKgmSQ1D0JoTDs0QXCdQvViZxHecMUXBhay3FZd33VhP3ImBpCYCAQLgAALShKmAAA3JAkTCiBkEUMLhJHJgePCgE9UBcimEQCB4GAqAQKVVh8sQbSVF0-SO2YnsKj7G8axdYMH0HHiuJWAYeVnfl7kpOYRFOTsRPpN41zObqhl5G5xn63gqQWFZemYNYGQlOAAEZfA5TtO25abermqZHiMUwaTeD5kxgMEVmkDQQXwW6ZGkfw9tSXw4Am9cNwCcrKr-PSM1PTk6svBrr0dQMWpDYbOA6xcuouQ6R2O6YFsG4F-MUCAIHwTUIZk2jXK9KNfU7fyAD0AH411G9Y3gO4kjvJdHZmpRiuVW+lNg27b5l8T9gSZigZrJAUhQwTnRS8CV7vURBZXBe7Xy+1Iqdp961a1uBNyAA) it shows the same error that just happens to be more helpful.

As well if you specify the `TAGS` entry the original error suggested [it gives the better message](https://www.typescriptlang.org/play?ts=4.5.0-dev.20211024#code/KYDwDg9gTgLgBMAdgVwLZwMrLJWwAmAYgJYA2MwUAhgEanADOcA3gFBwdwCCGGAkgHEAcgFERcALxwA5FQYNiAc0TBg0gDTtOAEQCqIgPrauAFXFTp+ZMG1UKGrRxNcBGSTJhVFDBwF9WrKC48EhocCJCJnwmAJoGsQAK4myc3O7SXA6pAELp2X4BQdDwMACeYMBwANLApQxciPgAalSk1gwGXO4pnADaWDjFBCTklLT0DAB0PPzCYgC6cABcLBVQDBCIfPgrKKg0lL6985qp-djBw2QU1HSMk3qGxmaLK8wMnrC2FCsfUMSIRTqBCNb7AX4wf6A3yafyBcDFOBlCrVWr1RotNqMAy5KQ9DjnQZ4IjXMZ3KYzQSiESvVaUDZbHZwPYHKBHE6OOCEy4k0a3CaTZyuWnMTyKbYQqFA5lUVDguB-AGKGGsOHw4JI8qVESIGDEMomLUmCAjG41OoNZqtdrdTm9CJRWLxGJJaa083oq1YjpcO0O6JxRIiSbZd1oy2Y9o41UBfDAADGpCoUEqADNkIh43rNkjGDBCBmADwiEwIEAURpMf1OoMAPgAFJyAWBkPAlvjUiC9QatSsS8DOalU6SoH3dfrSoaKsbTZQPRHrYx7SZ5oPOL4AJTLOBURClAIUD4FxD15hIbuT3vhSIB52urjA4d8t6+TdwAD077giiQlGI8bgAB3aAAGsmAYSB4E2ABCAJWGRSpMWIfAAGEIH2AE7GITYGELABZLgElrW0zhLOAATgEDaggVM4AIhJ5jec8JyneV+0o2pI3lKjShoujCOXE44CfG4GBWejBP8I4eL4+jV1jBMkxTYSMyzbDEGZTYBF-f542PeswCoUpSAgKgmSQ1D0JoTDs0QXCdQvViZxHecMUXBhay3FZd33VhP3ImBpCYCAQLgAALShKmAAA3JAkTCiBkEUMLhJHJgePCgE9UBcimEQCB4GAqAQKVVh8sQbSVF0-SO2YnsKj7G8axdYMH0HHiuJWAYeVnfl7kpOYRFOTsRPpN41zObqhl5G5xn63gqQWFZemYNYGQlOAAEZfA5TtO25abermqZHiMUwaTeD5kxgMEVmkDQQXwW6ZGkfw9tSXw4Am9cNwCcrKr-PSM1PTk6svBrr0dQMWpDYbOA6xcuouQ6R2O6YFsG4F-MUCAIHwTUIZk2jXK9KNfU7fyAD0AH411G9Y3gO4kjvJdHZmpRiuVW+lNg27b5l8T9gSZigZrJAUhQwTnRS8CV7vURBZXBe7Xy+1Iqdp961a1uBNyAA) saying that that wasn't valid to begin with since it doesn't satisfy the actual constraint.

```
Argument of type '{ entityType: ENTITY_TYPE.A; keyValue: SupportedFilterables.ASSIGNEE; filters: { assignee: { personId: number; }[]; tags: { tagId: string; name: string; }; }; }' is not assignable to parameter of type 'ValidCombinations<EntityTypeToFilterKeysAndValues>'.
  Types of property 'filters' are incompatible.
    Type '{ assignee: { personId: number; }[]; tags: { tagId: string; name: string; }; }' is not assignable to type 'KeysAndValues_A'.
      Object literal may only specify known properties, and '[SupportedFilterables.TAGS]' does not exist in type 'KeysAndValues_A'.(2345)
```

So what is actually happening is that typescript is making multiple passes over the data structure, on the first pass it checks if the `filters` is compatible with any of the valid targets for it, determines it isn't and gives the error which specifically mentions `KeysAndValues_B`, if it was a valid then on another pass it determines that actually the `filters` had to be set to `KeysAndValues_A` and may give a more accurate error message.  I'm not sure how actionable this would require basically detecting an error but checking a bit more details to possibly give a better error message.

**Comment by andrewbranch** [comment]
Likely a duplicate of https://github.com/microsoft/TypeScript/issues/46374. If you could make this repro a lot smaller and simpler, that would help make sure it is addressed along with #46374. I’m pretty sure you can get rid of a lot of the complexity here by inlining the type generated by `ValidCombinations`.

**Comment by Tomassito** [comment]
@tadhgmister So in the first pass of the compiler the generic value of Entity_Type is not yet resolved and taken into account, right?
When matching against the union type is it always the last type that will be used to produce error message ([playground](https://www.typescriptlang.org/play?ts=4.5.0-dev.20211024#code/KYDwDg9gTgLgBMAdgVwLZwMrLJWwAmAYgJYA2MwUAhgEanADOcA3gFBwdwCCGGAkgHEAcgFERcALxwA5FQYNiAc0TBg0gDTtOAEQCqIgPrauAFXFTp+ZMG1UKGrRxNcBGSTJhVFDB5zgANDgsQaVYAX1ZWUFx4JDQ4ESETPhMATQM0gAVxNj8ud2kuX04AIQKS4o4AYQKq0Iio8Gh4GABPMGA4AGlgVoYuRHwANSpSawYDfKlczgBtLBxmghJySlp6BgA6Hn5hMQBdOAAuFg6oBghEPnwTlFQaSjDZ-c0-eewY5bIKajpGTb0hmMZkOJ2YDE8sFsFBOEKgxEQinUCEG0OAsJg8MRYU0DWizTgbQ63V6-UGIzGjAMZWmjjg70WeCI3zWfy2O0EohEoNOlAuVxucDuDygTxedIZn2Zq1+G02zlcPOYnkU1wxWKRQqoqHRcDhCMUOPCkXxsEJ7U6PT6A2Go3GBhqtLeCylKx+63+HL23OOvPOlzVQrQIrFrzmLqW0vdbM2-iVxEFwseuMijRi5uJIkQMGIbRMFpMEDdlCtZNtlKYTrmiWSaQyqWy2x5pZtFPtXAlNZS6SyIk2JWbpNbdqpJU7SW79cbVUH1vJI4mVWNrHwwAAxqQqFBOgAzZCINc5y6ExgwQj7gA8IhMCBAFEGTC7dd7AD4ABR0hFgZDwI4zPwcEgOZ5haJzXsidJ+DuLJQGCswANa9HACJwIhrQQDuCTZrmrT5h0hbFlALbzhWszXvsPJZsBuEFkWMHEeW4xkSY+wIb0+wRABnBhAAlL6VCIK0kQUBC56IG+zBATheG6k+PYNn2XDItBMpgmEvFwAA9JpcCKEglDEGucAAO7QPBTAMJA8CXAAhKmRKdJkVCtKQEBUEQ0BUTh7j-vS17IYgqG9BhWHUTJBH0UOJHjPsYKQZwUkgR0YEmGGXFBa0ba6mhIVeUlwARTKDFZQwzEvHA8WcCpPwMHBaEBRluXYflhU-MVC5lZRzU0fhdFFVFjGMGVbGtBxFXpRwERPDlmF5T1BV9W1A0lfsqarhuW67vuh7EMeiCXAI+nwmuYlvmAzmue5JxOS5bkeVAc18ScAlCaw+2IIdKjHadvmJfNYETs+CnbGlHBoVlJwRkyhEeuyvCcmIyLabpEAQPgGbZcFmHtRWkzxdVfJxRNHCSpGMMxl6XKxfSzBnPygYAIxhOKxNaTppPQzBsOxvGgpM5V3HjcTvFAA))? 
Thank you for the explanation of the inner works - it helps to understand the 'whys'!
I am just surprised that the generic version of that function doesn't suffer from the same problem?

@andrewbranch 
As you suggested, I've inlined that type and removed some non-essential bits. ([playground](https://www.typescriptlang.org/play?ts=4.5.0-dev.20211024#code/KYOwrgtgBAymAO8D2AnALsAJgMQJYBsMUBDAI32AGcoBvAKCkagEEYYBJAcQDkBRXqAF4oAcmKVKuAOYhgwEQBoGTACIBVXgH0VzACoDhIzGGAriGRcsa7mnGENFpiUypYC+dOqEhRe3Xey6AJqawQAKAvRMLA4izJbRAEKxie6eaACe8MBQANLAGZTMIJgAasT4JpSazA5RTADacIioGDgERGQUlAB0rBw8-AC6UABctNkolEgg7Jjj4BCkwChuDUNK0U0IyOhYeIQrXVQ96lo6+iPjNJRO6GYY47couCBSClCgmA-AT2gvbzcSg8dEy2TyBSKJXKlSommSwnqjG2LT27UOJHIJ36XD4vCuExW01m8ygi2Wq3Wm0azV2bQOnSxvRsdgJNCcUjmfwB7zJxAgvygz1eUjWGzoILBOV4IDQuEyuiywF0SAZK3yhWKZQqVTqVigDT8AWCoSCET6BI1UO1sOqzH1hv8gRC4V4PUSlshWphVXhEs8mGAAGN8MQUDkAGZgEBBuUzKAYW7YaMAHl4uk+AA8MCVqEbnaaIgA+AAU+te8DAaDGSOin1l8oyiuy43TH310QjHRW1waAGsClBXlABxkkBHfA2FUqVWqUFbvTqqIbdEMCTK5dPsrPu-OvdCl5QV0N+wUhh460w3ABKMZQYggDLpKhoZMgEs0UCbptK1tOk2un0Hxdhiva0q0+y7scvQ4oM+LjA0NCTMSXJQAATG4543lAAD0OFQFIoArLgQZQAA7qgfbUJQyDVjMACEnigkqUBhMQGT4EgxA4KgG6NnqWzpkOIAjgU46Tt+zbKqqu4LgetpDNcHZMF+jZSa2ujUpeokZD6gqjuJfFbtJc5yTaVTHu22mMCBRCUL2o7CTphlTj+24yRiZl6Ue6ZrneRluSZsn7uZy6+aeGTnlAylXhKEUuZJM4eUQXmHkMTGBiGYaRtGsa4PGIAzJwREvEGb4lvA7Gcdx4xsRxXE8SgAW3uMD5PnQhUgMVsileVtaqcZf7Gi6ZpuswWmMKOenjOBaJztBfRsLi-ATTZu72bQuH4QAwkgEDwAQKxCmAUiEbc1BoAAFuYjjOK497hlAEC4BIIrkZdJGXUO1CFdWV05EG4jADFNI7BB6KMt0i0DHiikGkhRIzKhACMmGrUweGg6i9JQUyPQsjABKIRyqEiGAYC4Jgih8gK4wiMg8AiJhIOMG40XWaz15AA))

**Comment by tadhgmister** [comment]
[I cleaned up the example a little further](https://www.typescriptlang.org/play?ts=4.5.0-dev.20211024#code/KYOwrgtgBAogcgFQJIIJoH00AUZQN4BQUxUAglALxQDkp1ANESQEKU3MMEC+BBAliAAuwAE4AzAIYBjYFADSwAJ4BnUiAAmANQkAbMMGXpyhElAnLlfAOYhgsgFz4ADqOUB7EEnWPwEAEaiXADaALqMpmJubpFu5I54yoISIoIAIhLCjokiAlb0UKDq6ZlQ2blcjDz8QqKSMvJKqhraegborCYk5pY2dlDxLiLunt5QvgEiwWFMxH7JcyKs8UlWXlmCOSB5YxIQwOubVlOVvIKKLrBCfGcI58AIbgBifDrCIgoqalq6+spsncQgvBkGhMKgcAA6UghfoNT7NH5tUgzKBAxAoDDYGAQ5gwxwfJrfVqGZjcapvOqyJAgJxgQQPR5gEBSAA8MAQBQAHsINH9gRiwTgAHwAgpXG53RzslFiF5vZRS8WKW4uBly0QEr4tX5AhAhMkEdTAKQ6ZKyMRMqSCPgeKDCRKMkBsjnAbmFPno0FYoUACgEtME-WpAYZludQoAlLCAG5uPjqAgAekTUCkbhEImNgh0igKGfTf0EAAsMjQYjE6FA+H9bMAjeoqyAoLLXq4CPbBI6fXhQNaJS4pZ7MeDsaR8i23vFutZa44gnhBsM1lAAExcEJcLhR5NQKygUR8KRQADu6YA1n9lE43IGPABCXjtu5QACqIBtIAA8mJSDodMG6T+KhRV1RsoDPJRokuPtlTuNVW3eRotURZQQkcAD6SeMN2SFbggggxQoJgJUVXuJ51UQ+EiV+fVDWNU1M2bS1rVtEAPAAcX3HIpC7JwJEUHQ3AkUY3w-b9f3-GlAKjRxY3jXg2JATjbG4rtRV7a5YIHWAh0FUdGB3UiaDwMwLBnOx4igRcPGXcZRAAbigKZHK4agqxrG9TJ6EAJD8HRZEENw7WfahNQRYkjCgAAfOFCW1NoOAhJMU2ILARDcQYzhoBYFg4dyoAgatLC2MCzguagTOnXp9nwKzXBs0Y7JEFzQhcty-DpKBMwARzAPhMwbARgvKsLqIS6gkp3UxppIAA9eaFtmqAAikCQwGUWQoIAAw0-sav5L0RyhLa7SLat3JAahAwkKBo10eMrPSzLFHyY9ZFWy7A2AaNQGWgLn0WxaUWICdXHiFEd1mgB+YHpqq2dUQXeqRkcABGddwhmkgd2pYQ-2rUANtTdNMytHMoA8cnojEVwyyiCs3KLURgFh0wuCgFEtwIRTlIPHimW7FFdq0-bdKxKFMYiCiFXwVmujM6q5yRoYGrRjG5Z3YT1FyU7ZCcJ7RCy4tZAEa4+F0PN0pEUowCsPdEjrMY3GPUp+L+a58oEO6dHkwy7j+KD9Yyw3c2oUGhjcs1GzTCA+OtPzgEmlKoCMiqvPMmqTOslGxkgCYWpCRycuSZgfG+0RWqctzzrY66FZ8hO7SCsrZFCpDwt+IwJrl2Z5hLud9Sxkh2c5iMgA)

The reason it works in the generic case is that the type of `filters` depends on resolving the generic first so it never gets the chance to validate against a wider union.
